### ISSUE TITLE
Query: Recognize additional join patterns

### ISSUE THREAD

**Issue by maumar** [aspnet/EntityFrameworkCore#3263]
In EF6 we support easy patterns to create join queries:

```
var query = from sna in ctx.Sna
            from foo in ctx.Foo.Where(foo => foo.Id == sna.Id)
            select new { sna, foo };
```

translates to:

```
SELECT
    [Extent1].[Id] AS [Id],
    [Extent2].[Id] AS [Id1]
    FROM  [dbo].[Sna] AS [Extent1]
    INNER JOIN [dbo].[Foo] AS [Extent2] ON [Extent2].[Id] = [Extent1].[Id]
```

```
var query = from sna in ctx.Sna
            from foo in ctx.Foo.Where(foo => foo.Id == sna.Id).DefaultIfEmpty()
            select new { sna, foo };
```

translates to:

```
SELECT
    [Extent1].[Id] AS [Id],
    [Extent2].[Id] AS [Id1]
    FROM  [dbo].[Sna] AS [Extent1]
    LEFT OUTER JOIN [dbo].[Foo] AS [Extent2] ON [Extent2].[Id] = [Extent1].[Id]
```

This also allows making joins using different operators than equal:

```
var query = from sna in ctx.Sna
            from foo in ctx.Foo.Where(foo => foo.Id != sna.Id).DefaultIfEmpty()
            select new { sna, foo };
```

translates to:

```
SELECT
    [Extent1].[Id] AS [Id],
    [Extent2].[Id] AS [Id1]
    FROM  [dbo].[Sna] AS [Extent1]
    LEFT OUTER JOIN [dbo].[Foo] AS [Extent2] ON [Extent2].[Id] <> [Extent1].[Id]
```

We should consider those for EF Core as well.

**Comment by MovGP0** [comment]
There is an additional pattern. Consider: 

```csharp
from foo in Context.Foos
join bar in Context.Bars on foo.BarId equals bar.BarId into fooBars
from fooBar in fooBars.DefaultIfEmpty()
select new { fooBar.FooId, fooBar.BarId }
```

```sql
SELECT [t2].[value] AS [BarId], [t2].[value2] AS [FooId]
FROM (
    SELECT [t1].[BarId] AS [value], [t1].[FooId] AS [value2]
    FROM [Foo] AS [t0]
    LEFT OUTER JOIN [Bar] AS [t1] ON ([t0].[FooId]) = [t1].[FooId]
    ) AS [t2]
```

vs. 

```csharp
from foo in Context.Foos
from bar in Context.Bars.DefaultIfEmpty()
select new { foo.FooId, bar.BarId }
```

```sql
SELECT [t2].[value] AS [FooId], [t2].[FooId]
FROM (
    SELECT [t1].[FooId] AS [value], [t0].[FooId]
    FROM [Foo] AS [t0]
    LEFT OUTER JOIN [Bar] AS [t1] ON 1=1 
    ) AS [t2]
```

**Comment by smitpatel** [comment]
We already translate patterns to generate join in 3.0. We don't process non-equal conditions yet. Filed #18871 to track it. Closing this as fixed in 3.0

**Comment by smitpatel** [comment]
Fixed in #17077
